website/python/brewmonitor/storage/tables.py

```diff
@@ -0,0 +1,710 @@
+    min_battery = attr.ib(type=float, default=None, metadata={'sql': '{name} real'})
+
+    last_active = attr.ib(
+        type=datetime,
+        default=None,
+        metadata={'db_factory': datetime_row_factory('last_active')},
+    )  # in SQL we get that from the datapoints
+    last_battery = attr.ib(type=float, default=None)  # in SQL we get that from the datapoints
+    linked_project = attr.ib(type=int, default=None)  # Assuming a sensor is only attached to 1 project
+
+    @classmethod
+    def additional_sql_fields(cls) -> List[str]:
+        return super(Sensor, cls).additional_sql_fields() + [
+            'battery float',
+            'foreign key(owner) references User(id)',
+        ]
+
+    @classmethod
+    def get_all(cls, db_conn: SQLConnection, **kwargs) -> List["Sensor"]:
+        cursor = db_conn.execute(
+            '''
+            select id, name, secret, max_battery, min_battery, 
+                (select battery from Datapoint where sensor_id = Sensor.id order by timestamp desc limit 1) as last_battery,
+                (select timestamp from Datapoint where sensor_id = Sensor.id order by timestamp desc limit 1) as last_active,
+                (select username from User where id = Sensor.owner limit 1) as owner,
+                (select id from Project where active_sensor = Sensor.id) as linked_project
+            from Sensor
+            order by id desc;
+            '''
+        )
+        cursor.row_factory = cls.row_factory
+        return cursor.fetchall()
+
+    @classmethod
+    def find(cls, db_conn: SQLConnection, sensor_id: int = Required) -> Optional["Sensor"]:
+        if sensor_id is None:
+            raise ValueError('sensor_id is required')
+        sens_cursor = db_conn.execute(
+            '''
+            select id, name, secret, max_battery, min_battery, 
+                (select battery from Datapoint where sensor_id = Sensor.id order by timestamp desc limit 1) as last_battery,
+                (select timestamp from Datapoint where sensor_id = Sensor.id order by timestamp desc limit 1) as last_active,
+                (select username from User where id = Sensor.owner limit 1) as owner,
+                (select id from Project where active_sensor = Sensor.id) as linked_project
+            from Sensor where id = ?;
+            ''',
+            (sensor_id,)
+        )
+        sens_cursor.row_factory = cls.row_factory
+        return sens_cursor.fetchone()
+
+    @classmethod
+    def create(
+        cls,
+        db_conn: SQLConnection,
+        name: str = Required,
+        secret: str = Required,
+        owner: User = Required,
+    ) -> "Sensor":
+        if name is None or secret is None or owner is None:
+            raise ValueError('All arguments are required')
+
+        cursor = db_conn.cursor()
+        cursor.execute(
+            '''
+            insert into Sensor (name, secret, owner, max_battery, min_battery) values (?, ?, ?, 4.0, 2.0);
```

Comment:
Not the best way to do defaults, we should use the attr attribute to store the default and more importantly you HAVE to pass these values bellow when creating the object!